scipy/optimize/_numdiff.py

```diff
@@ -551,32 +551,43 @@ def _dense_difference(fun, x0, f0, h, use_one_sided, method):
     h_vecs = np.diag(h)
 
     for i in range(h.size):
+        df = np.zeros_like(f0)
+
         if method == '2-point':
             x = x0 + h_vecs[i]
             dx = x[i] - x0[i]  # Recompute dx as exactly representable number.
-            df = fun(x) - f0
+            if np.count_nonzero(dx):
```

Comment:
For my understanding: `dx` is a float here, why not simply write `dx != 0`?

Reply:
`dx` can also be an array

Reply:
Not for the sake of proving you wrong, but after all ifs `df / dx` is computed, which corresponds to a column in Jacobian. And just by indexing `dx = x[i] - x0[i]` it really looks like a float. Maybe I'm missing something.

It's not terribly important, but if it is a float, then maybe it's better not to use `np.count_nonzero`.

Reply:
`h_vecs` can be a array size > 1.

Reply:
Somewhat tricky logic is involved here.

`h_vecs = np.diag(h)` and `x = x0 + h_vecs[i]` - meaning that we add deltas of the form `[0, 0, h[2], 0, 0, 0]`, etc. And after that we compute scalar `dx = x[i] - x0[i]`, which will be very close to `h[i]` but more accurate considering floating-point arithmetic (the comment `# Recompute dx as exactly representable number.`).

At the end `J_transposed[i] = df / dx` -- i-th column of the Jacobian is vector `df` divided by scalar `dx`.

Reply:
Using `count_nonzero` means that we can avoid function evaluations for which we know all entries in `dx` are zero. I should've said in my previous comment that `dx` can have size > 1, it's not necessarily a scalar.

Reply:
Using `np.count_nonzero` or `if dx != 0` allows one to avoid a function evaluation if the step size is zero, which it will be for equal bounds.

Reply:
Yeah, I see. But `np.count_nonzero` seems out of place a little bit, considering that `dx` is float (rechecked by adding additional `assert np.array(dx).ndim == 0` and running all the tests).

Reply:
I agree with your arguments @nmayorov, made those changes.